service/matching/tasklist/isolation_balancer.go

```diff
@@ -0,0 +1,410 @@
+	slices.SortFunc(toRemove, byValueDescending(groupSizePerPartition))
+	for _, group := range toRemove {
+		partitionID, ok := findMaxWithGroup(group, groupSizePerPartition, groupsByPartitionID)
+		if ok {
+			delete(groupsByPartitionID[partitionID], group)
+		}
+	}
+}
+
+func addPartitions(toAdd []string, groupSizePerPartition map[string]float64, groupsByPartitionID map[int]map[string]any) {
+	// Use greedy number partitioning to add the group to the smallest partition where it isn't already present. This
+	// is an approximate solution. Sorting the input by size in a descending manner should result in a more even
+	// distribution.
+	slices.SortFunc(toAdd, byValueDescending(groupSizePerPartition))
+	for _, group := range toAdd {
+		partitionID, ok := findMinWithoutGroup(group, groupSizePerPartition, groupsByPartitionID)
+		if ok {
+			groupsByPartitionID[partitionID][group] = true
+		}
+	}
+}
+
+func ensureMinimumGroupsPerPartition(minGroups int, groupSizePerPartition map[string]float64, groupsByPartitionID map[int]map[string]any) bool {
+	// Every partition needs to be assigned groups that have pollers, otherwise we're unable to process tasks on those
+	// partitions.
+	//
+	// Sort the partitions by the number of isolation groups assigned to them, and then use a two-pointers approach
+	// to iteratively take groups from partitions that have the most and move them to a partition that has the least.
+	// calculatePartitionGoal enforces that there are at least "getMinPartitionsPerGroup" instances of each group,
+	// which in turn ensures that every partition can be assigned "getMinGroupsPerPartition" by only moving groups
+	// around.
+	//
+	// The specific groups that we move between partitions isn't particularly significant, since the size of each group
+	// within a partition has an inherent upper-bound before we scale up. Ideally we keep partitions relatively close
+	// in size, so we use a simple heuristic of minimizing the difference in total size between the two partitions that
+	// we're considering.
+	changed := false
+	partitionIDsByGroupCount := maps.Keys(groupsByPartitionID)
+	slices.SortFunc(partitionIDsByGroupCount, func(a, b int) int {
+		aLen := len(groupsByPartitionID[a])
+		bLen := len(groupsByPartitionID[b])
+		if aLen < bLen {
+			return -1
+		} else if aLen == bLen {
+			// Fall back to size ascending. Larger is later to prioritize moving groups from those partitions
+			aSize := getPartitionSize(groupSizePerPartition, groupsByPartitionID[a])
+			bSize := getPartitionSize(groupSizePerPartition, groupsByPartitionID[b])
+			if aSize < bSize {
+				return -1
+			} else if aSize == bSize {
+				return 0
+			} else {
+				return 1
+			}
+		} else {
+			return 1
+		}
+	})
+	lo, hi := 0, len(partitionIDsByGroupCount)-1
+	for lo < hi {
+		loPartition := groupsByPartitionID[partitionIDsByGroupCount[lo]]
+		hiPartition := groupsByPartitionID[partitionIDsByGroupCount[hi]]
+		if len(loPartition) >= minGroups {
```

Comment:
if the current partition has enough groups, shouldn't the next partition also have enough groups because it's sorted in ascending order?

Reply:
Discussed offline with some examples. This condition is met when the current `lo` is satisfied, but we might need to still process the next one. It's because we don't change lo or hi after moving groups between them.